civis/io/_files.py

```diff
@@ -256,10 +257,13 @@ def _file_to_civis(buf, name, api_key=None, client=None, **kwargs):
         log.warning('Could not determine file size; defaulting to '
                     'single post. Files over 5GB will fail.')
 
-    # determine if file-like object is seekable
+    # determine if file-like object will support multipart upload
+    multipart_upload = False
     try:
         buf.seek(buf.tell())
         is_seekable = True
+        if hasattr(buf, 'name') and hasattr(buf, 'mode'):
+            multipart_upload = True
```

Comment:
Could you use only these hasattr checks instead of trying to seek? Shouldn't every buffer in this function be seekable?

Reply:
Theoretically I would get right of this whole try block and the use of `is_seekable` and `multipart_upload` variables because in `file_to_civis` we are checking if the buffer `isinstance(buf, io.BufferedReader)` which should ensure that it is seekable and has these attributes.

I have come across file like objects that have a seek method, but fail when you call it, which makes me sad.

@stephen-hoover Let me know what you think.

Reply:
I just realized that `is_seekable` shouldn't be necessary any longer because we write files to disk if `.tell() != 0`. So we should be able to remove the `.seek` call, correct? 

We'd still be unable to do a multipart upload if we can't find a `name`.  Do you think it's better to handle that by writing to disk and doing a multipart upload, or by reverting to `_single_upload` as we are now? I think the `_single_upload` is better, but I could easily be convinced to write to disk and treat everything the same.

Reply:
I think we still need the `.seek` for retries and to seek to different positions on the buffer for multipart upload.

I struggled a bit with deciding what checks should go in `file_to_civis` versus `_file_to_civis`. The `isinstance(buf, io.BufferedReader)` check is designed to catch in memory buffers. If that check is satisfied, then the file like object should have a `name`, unless someone did something silly, and it should be seekable no? As long as `file_to_civis` fails and returns the underlying error should we try to handle edge cases like `is_seekable` or `name`?

Regardless, I feel like the `isinstance` check is us changing how the `file_to_civis` function works, but making the function backwards compatible so the `file_to_civis` location makes sense.

The `is_seekable` check is for determining if we can do retries or multipart upload (which requires seeking), which impacts what branch of the code is executed so it feels like that belongs in `_file_to_civis`.

Checks on `buf.tell` and `buf.name` are guardrails to handle possible edge cases.

Reply:
It looks like the `io.BufferedReader` is guaranteed to have `seek`, but not guaranteed to have `name`.

The multipart upload only seeks on the buffer we open ourselves from the file on disk, correct? It doesn't need the input buffer to be seekable. You're right about retries, though. But we only need to seek for retries if we're uploading directly from an in-memory buffer. So in which cases do we want to upload from the in-memory buffer, and in which cases do we read from disk and do a multipart upload?

Reply:
Thats correct, for multipart upload we seek on the buffer we open ourselves.

I don't think we're allowing uploads from an in-memory buffer, since we re-write in-memory buffers to disk if users pass them in `file_to_civis` before calling `_file_to_civis`. Given that we could arguable strip out `seekable` check altogether couldn't we? Although, I don't know if its good to keep it for visibility.

Reply:
Since we're making strong assumptions about what goes into this function already, I think we should leverage that to make the code as simple as possible.